File: file_processing/activities.py
import socket
import ssl
import tempfile
import urllib.request
from dataclasses import dataclass
from pathlib import Path

from temporalio import activity

# The shared task queue any worker can pick up.
DEFAULT_TASK_QUEUE = "file-processing-task-queue"


@dataclass
class DownloadResult:
    """Carries the host-specific task queue and the local path of the downloaded file."""

    host_task_queue: str
    local_path: str


@dataclass
class UploadInput:
    """Input for the upload activity."""

    local_path: str
    destination_url: str


@activity.defn
async def download(source_url: str) -> DownloadResult:
    """
    Download a file from source_url to a temporary local file.

    Returns the path and the task queue name for *this* worker so the workflow
    can pin the next activities to the same host.
    """
    activity.logger.info(f"Downloading {source_url}")

    # NOTE: SSL verification is disabled here for convenience and should not be used in production.
    # In production, remove the ssl_context argument to enforce certificate validation.
    ssl_context = ssl.create_default_context()
    ssl_context.check_hostname = False
    ssl_context.verify_mode = ssl.CERT_NONE

    with urllib.request.urlopen(source_url, context=ssl_context) as response:
        content = response.read()

    # Write to a named temp file that persists after this function returns.
    tmp = tempfile.NamedTemporaryFile(delete=False, suffix=".tmp")
    tmp.write(content)
    tmp.close()

    activity.logger.info(f"Downloaded {source_url} → {tmp.name}")

    # Tell the workflow which task queue (i.e. which host) now holds the file.
    host_task_queue = socket.gethostname()
    return DownloadResult(host_task_queue=host_task_queue, local_path=tmp.name)


@activity.defn
async def process(local_path: str) -> str:
    """
    Process the downloaded file.

    In a real scenario this might compress, transcode, or analyse the file.
    Here we simply count its bytes as a stand-in for real work and return the
    (unchanged) path so the upload activity knows where to find it.
    """
    path = Path(local_path)
    size = path.stat().st_size
    activity.logger.info(f"Processing {local_path} ({size} bytes)")

    # Simulate processing by appending a small metadata footer.
    with open(local_path, "ab") as f:
        f.write(b"\n-- processed --\n")

    activity.logger.info(f"Processing complete: {local_path}")
    return local_path


@activity.defn
async def upload(input: UploadInput) -> None:
    """
    Upload the processed file to the destination.

    Here we simulate the upload by logging the action; swap in a real HTTP PUT
    or cloud SDK call for production use.
    """
    path = Path(input.local_path)
    size = path.stat().st_size
    activity.logger.info(f"Uploading {input.local_path} ({size} bytes) → {input.destination_url}")

    # Simulate upload delay / work.
    # (Replace with actual upload logic, e.g. boto3 / requests.)
    activity.logger.info(f"Upload complete: {input.local_path} → {input.destination_url}")

    # Clean up the temporary file now that it has been "uploaded".
    path.unlink(missing_ok=True)

File: file_processing_ts/src/activities.ts
import { log } from '@temporalio/activity';
import * as fs from 'fs/promises';
import * as os from 'os';
import * as path from 'path';

export interface DownloadResult {
  /** The host-specific task queue (hostname) where the file was downloaded. */
  hostTaskQueue: string;
  /** The local path of the downloaded file. */
  localPath: string;
}

export interface UploadInput {
  localPath: string;
  destinationUrl: string;
}

/**
 * Activities that can run on any worker (no local file dependency).
 */
export function createDefaultActivities() {
  return {
    /**
     * Download a file from sourceUrl to a temporary local file.
     *
     * Returns the local path and this worker's hostname so the workflow can
     * pin the next activities to the same host.
     */
    async download(sourceUrl: string): Promise<DownloadResult> {
      log.info('Downloading file', { sourceUrl });

      const response = await fetch(sourceUrl);
      const content = await response.text();

      const localPath = path.join(os.tmpdir(), `temporal-${Date.now()}.tmp`);
      await fs.writeFile(localPath, content, 'utf8');

      log.info('Download complete', { sourceUrl, localPath });

      // Return the hostname so the workflow knows which task queue (host) to
      // target for the process and upload activities.
      return { hostTaskQueue: os.hostname(), localPath };
    },
  };
}

/**
 * Activities that must run on the same host that downloaded the file,
 * because they operate on the local filesystem.
 */
export function createHostActivities() {
  return {
    /**
     * Process the downloaded file.
     *
     * In a real scenario this might compress, transcode, or analyse the file.
     * Here we append a small metadata footer as a stand-in for real work.
     */
    async process(localPath: string): Promise<string> {
      const stat = await fs.stat(localPath);
      log.info('Processing file', { localPath, bytes: stat.size });

      await fs.appendFile(localPath, '\n-- processed --\n', 'utf8');

      log.info('Processing complete', { localPath });
      return localPath;
    },

    /**
     * Upload the processed file to the destination.
     *
     * Simulated here; swap in a real HTTP PUT or cloud SDK call for production.
     */
    async upload({ localPath, destinationUrl }: UploadInput): Promise<void> {
      const stat = await fs.stat(localPath);
      log.info('Uploading file', { localPath, bytes: stat.size, destinationUrl });

      // Replace with real upload logic (e.g. fetch PUT, AWS S3 putObject, etc.)
      log.info('Upload complete', { localPath, destinationUrl });

      // Clean up the temp file now that it has been "uploaded".
      await fs.rm(localPath, { force: true });
    },
  };
}
